api_core/google/api_core/grpc_helpers.py

```diff
@@ -136,3 +138,178 @@ def create_channel(target, credentials=None, scopes=None, **kwargs):
 
     return google.auth.transport.grpc.secure_authorized_channel(
         credentials, request, target, **kwargs)
+
+
+_ChannelCall = collections.namedtuple(
+    '_ChannelCall', ['request', 'timeout', 'metadata', 'credentials'])
+
+
+class _CallableStub(object):
+    """Stub for the grpc.*MultiCallable interfaces."""
+
+    def __init__(self, method, channel):
+        self._method = method
+        self._channel = channel
+        self.response = None
+        """Union[protobuf.Message, Callable[protobuf.Message], exception]:
+        The response to give when invoking this callable. If this is a
+        callable, it will be invoked with the request protobuf. If it's an
+        exception, the exception will be raised when this is invoked.
+        """
+        self.responses = None
+        """Iterator[protobuf.Message]: A iterator of responses. If specified,
+        self.response will be populated on each invocation by calling
+        ``next(self.responses)``."""
+        self.requests = []
+        """List[protobuf.Message]: All requests sent to this callable."""
+        self.calls = []
+        """List[Tuple]: All invocations of this callable. Each tuple is the
+        request, timeout, metadata, and credentials."""
+
+    def __call__(self, request, timeout=None, metadata=None, credentials=None):
+        self._channel._requests.append((self._method, request))
+        self.requests.append(request)
+        self.calls.append(
+            _ChannelCall(request, timeout, metadata, credentials))
+
+        if self.responses:
+            self.response = next(self.responses)
+
+        if callable(self.response):
+            return self.response(request)
+
+        if isinstance(self.response, Exception):
+            raise self.response
+
+        if self.response is not None:
+            return self.response
+
+        raise ValueError(
```

Comment:
Is there a valid use case where someone creates a stub with no responses?

When I wrote an equivalent library for HTTP, I required pre-registration of endpoints you planned to use (so I could error if you used something else), but you already have that. That said, I am not sure there _is_ a valid situation in which GRPC succeeds (on a unary-unary request) and provides no response.

Reply:
No, an RPC *must* return a response (that response can totally be `empty_pb2.Empty()`, see the tests for `operations_client`.) This case is the equivalent of your client-under-test making an RPC that you do not anticipate so it should throw an error.